src/backend/env/auction.rs

```diff
@@ -0,0 +1,393 @@
+    pub e8s_per_token: u64,
+    timestamp: Time,
+}
+
+impl PartialOrd for Bid {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Bid {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        if self.e8s_per_token != other.e8s_per_token {
+            // prioritize higher bids
+            return self.e8s_per_token.cmp(&other.e8s_per_token);
+        }
+        if self.timestamp != other.timestamp {
+            // prioritize older bids
+            return other.timestamp.cmp(&self.timestamp);
+        }
+        if self.amount != other.amount {
+            // prioritize larger bids
+            return self.amount.cmp(&other.amount);
+        }
+        self.user.cmp(&other.user)
+    }
+}
+
+#[derive(Default, Serialize, Deserialize)]
+pub struct Auction {
+    pub amount: Token,
+    bids: BTreeSet<Bid>,
+}
+
+impl Auction {
+    /// Checks if there is a sell out.
+    pub fn sell_out(&self) -> bool {
+        self.bids.iter().map(|bid| bid.amount).sum::<u64>() >= self.amount
+    }
+
+    /// Returns the highest bids if they form a sell out.
+    pub fn get_bids(&mut self) -> Vec<Bid> {
+        if !self.sell_out() {
+            return Default::default();
+        }
+        let mut amount = self.amount;
+        let mut bids = Vec::default();
+
+        while amount > 0 {
+            let Some(bid) = self.bids.pop_last() else {
+                break;
+            };
+            let eff_amount = bid.amount.min(amount);
+            bids.push(Bid {
+                user: bid.user,
+                amount: eff_amount,
+                e8s_per_token: bid.e8s_per_token,
+                timestamp: bid.timestamp,
+            });
+            // if the bid was larger than tokens left, create a leftover bid
+            if eff_amount < bid.amount {
+                self.bids.insert(Bid {
+                    user: bid.user,
+                    amount: bid.amount - eff_amount,
+                    e8s_per_token: bid.e8s_per_token,
+                    timestamp: bid.timestamp,
+                });
+            }
+            amount = amount.saturating_sub(eff_amount);
+        }
+
+        bids
+    }
+}
+
+/// Cancels user's bid and returns the funds to user wallet.
+pub async fn cancel_bid(principal: Principal) -> Result<u64, String> {
+    let bid = mutate(|state| remove_bid(state, principal))?;
+
+    let user_account = read(|state| {
+        let user_principal = state
+            .principal_to_user(principal)
+            .expect("no user found")
+            .principal;
+        AccountIdentifier::new(&user_principal, &DEFAULT_SUBACCOUNT)
+    });
+
+    let funds = bid
+        .amount
+        .checked_mul(bid.e8s_per_token)
+        .expect("overflow")
+        .checked_sub(DEFAULT_FEE.e8s())
+        .expect("nothing to refund");
+
+    invoices::transfer(
+        user_account,
+        Tokens::from_e8s(funds),
+        Memo(727),
```

Comment:
Is this 727 meaningful? or the 717 and 737 used as memos for the other transactions. Is it worth defining these in one place as named constants (along with any other memo currently used elsewhere - 999, 0x50555054, 828282)?

Reply:
They have no meaning right now, I just use different numbers to be able to distinguish txs in the future if needed (ever). Then yes we could have a docs or something like that.